### ISSUE TITLE
archive_file changes file permissions on Windows

### ISSUE THREAD

**Issue by KyleKotowick** [hashicorp/terraform-provider-archive#58]
This is a re-opening of issue #10 .

This issue is still occurring on Windows. Within the ZIP file created by archive_file when running Terraform v0.12.18 with AWS provider v2.43 on Windows 10, the contained file has 666 permissions (no execute). Running the exact same Terraform plan with the same version on Linux results in the contained file having 777 permissions.

**Comment by RichardBradley** [comment]
I have a similar or the same issue.
I am creating a zip file for a Lambda with code like this:

```
data "archive_file" "my_lambda_zip" {
  type        = "zip"
  output_path = "${path.root}/.terraform/tmp/my_lambda.zip"
  source_file = "${path.module}/my-src.js"
}
```

The "my-src.js" file is a regular file in git:
```
$ git ls-files -s my-src.js
100644 8ef1d00576701a5cbd7c07c0f5496131b4554b43 0       my-src.js
```

On unix, the output of the "archive_file" is a zip containing a 644 / "-rw-r--r--" file:
```
$ zipinfo my_lambda.zip
Archive:  my_lambda.zip
Zip file size: 1064 bytes, number of entries: 1
-rw-r--r--  2.0 unx     2001 bl defN 49-Jan-01 00:00 my-src.js
1 file, 2001 bytes uncompressed, 876 bytes compressed:  56.2%
```

On windows, the file permissions are different:

```
zipinfo my_lambda.zip
Archive:  my_lambda.zip
Zip file size: 1064 bytes, number of entries: 1
-rw-rw-rw-  2.0 unx     2001 bl defN 49-Jan-01 00:00 my-src.js
1 file, 2001 bytes uncompressed, 876 bytes compressed:  56.2%
```

This causes Terraform to unnecessarily mark the Lambda as changed, as I am using "filebase64sha256" on the zip, which differs by this one byte.

How can I persuade archive_file to generate a 644/"-rw-r--r--" file on Windows (or indeed a 666/"-rw-rw-rw-" file on Unix)?

Thanks

**Comment by RichardBradley** [comment]
It looks like the code that handles this reads the file perms using Go's "os.stat" at [archiver.go:28](https://github.com/hashicorp/terraform-provider-archive/blob/8730bafba6d024c9ef5e3665c511cce8a5f205ff/internal/provider/archiver.go#L28) and hands that file mode off directly to Go's zip.FileInfoHeader at [zip_archiver.go:56](https://github.com/hashicorp/terraform-provider-archive/blob/main/internal/provider/zip_archiver.go#L56).

According [to this blog](https://medium.com/@MichalPristas/go-and-file-perms-on-windows-3c944d55dd44), Go will always report either 666 or 444 file mode.

So without a windows-specific workaround to archive_file, I will need to arrange for my file to be 666 on unix to make my builds consistent cross-platform.

**Comment by RichardBradley** [comment]
It looks like git won't allow me to make the file 666 mode, only 644 or 755 ([cite](https://medium.com/@tahteche/how-git-treats-changes-in-file-permissions-f71874ca239d)).
So I'm going to struggle to get this to behave the same on unix and windows without some kind of helper script

**Comment by kmoe** [comment]
The workaround in #90 has been released in terraform-provider-archive v2.2.0. If output_file_mode does not solve your problem, please comment on this issue or open a new one.

**Comment by RichardBradley** [comment]
Thanks, that workaround fixes this for me!
